sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs

```diff
@@ -0,0 +1,439 @@
+        /// </summary>
+        ///
+        private readonly string EventHubName;
+
+        /// <summary>
+        ///   The name of the consumer group this event processor is associated with.  Events will be
+        ///   read only in the context of this group.
+        /// </summary>
+        ///
+        private readonly string ConsumerGroup;
+
+        /// <summary>
+        ///   A unique name used to identify this event processor.
+        /// </summary>
+        ///
+        private readonly string Identifier;
+
+        /// <summary>
+        ///   The minimum amount of time for an ownership to be considered expired without further updates.
+        /// </summary>
+        ///
+        private TimeSpan OwnershipExpiration;
+
+        public IEnumerable<string> OwnedPartitionIds => InstanceOwnership.Keys;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PartitionLoadBalancer"/> class.
+        /// </summary>
+        ///
+        public PartitionLoadBalancer(PartitionManager storageManager,
+                                      string identifier,
+                                      string consumerGroup,
+                                      string fullyQualifiedNamespace,
+                                      string eventHubName,
+                                      TimeSpan ownershipExpiration)
+        {
+            Argument.AssertNotNull(storageManager, nameof(storageManager));
+            Argument.AssertNotNullOrEmpty(identifier, nameof(identifier));
+            Argument.AssertNotNullOrEmpty(consumerGroup, nameof(consumerGroup));
+            Argument.AssertNotNullOrEmpty(fullyQualifiedNamespace, nameof(fullyQualifiedNamespace));
+            Argument.AssertNotNullOrEmpty(eventHubName, nameof(eventHubName));
+
+            StorageManager = storageManager;
+            Identifier = identifier;
+            FullyQualifiedNamespace = fullyQualifiedNamespace;
+            EventHubName = eventHubName;
+            ConsumerGroup = consumerGroup;
+            OwnershipExpiration = ownershipExpiration;
+
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="PartitionLoadBalancer"/> class.
+        /// </summary>
+        ///
+        protected PartitionLoadBalancer()
+        {
+        }
+
+        /// <summary>
+        ///   Performs load balancing between multiple <see cref="EventProcessorClient" /> instances, claiming others' partitions to enforce
+        ///   a more equal distribution when necessary.  It also manages its own partition processing tasks and ownership.
+        /// </summary>
+        ///
+        /// <param name="partitionIds"></param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        internal virtual async ValueTask<PartitionOwnership> RunAsync(string[] partitionIds, CancellationToken cancellationToken)
```

Comment:
May as well designate all `internal` members public, as the class visibility constrains.   Not that it is impactful, but that's the pattern that has been used.